vendor/bat-native-confirmations/src/bat/confirmations/internal/balance.cc

```diff
@@ -0,0 +1,379 @@
+  double estimated_pending_rewards = 0.0;
+
+  for (const auto& payment : payments_) {
+    estimated_pending_rewards += payment.balance;
+  }
+
+  estimated_pending_rewards -= ad_grants_amount_;
+  if (estimated_pending_rewards < 0.0) {
+    estimated_pending_rewards = 0.0;
+  }
+
+  return estimated_pending_rewards;
+}
+
+uint64_t Balance::CalculateNextPaymentDateInSeconds(
+    uint64_t next_token_redemption_date_in_seconds) const {
+  auto now = base::Time::Now();
+  base::Time::Exploded now_exploded;
+  now.LocalExplode(&now_exploded);
+
+  int month = now_exploded.month;
+
+  if (now_exploded.day_of_month <= kNextPaymentDay) {
+    if (HasPendingBalanceForLastMonth()) {
+      // If last month has a pending balance, then the next payment date will
+      // occur this month
+    } else {
+      // If last month does not have a pending balance, then the next payment
+      // date will occur next month
+      month++;
+    }
+  } else {
+    if (HasPendingBalanceForThisMonth()) {
+      // If this month has a pending balance, then the next payment date will
+      // occur next month
+      month++;
+    } else {
+      auto next_token_redemption_date =
+          base::Time::FromDoubleT(next_token_redemption_date_in_seconds);
+
+      base::Time::Exploded next_token_redemption_date_exploded;
+      next_token_redemption_date.LocalExplode(
+          &next_token_redemption_date_exploded);
+
+      if (next_token_redemption_date_exploded.month == month) {
+        // If this month does not have a pending balance and our next token
+        // redemption date is this month, then the next payment date will occur
+        // next month
+        month++;
+      } else {
+        // If this month does not have a pending balance and our next token
+        // redemption date is next month, then the next payment date will occur
+        // the month after next
+        month += 2;
+      }
+    }
+  }
+
+  int year = now_exploded.year;
+
+  if (month > 12) {
+    month -= 12;
+    year++;
+  }
+
+  std::string date = std::to_string(year);
+  date += "-";
+  date += std::to_string(month);
+  date += "-";
+  date += std::to_string(kNextPaymentDay);
+
+  base::Time next_payment_date;
+  bool success = base::Time::FromString(date.c_str(), &next_payment_date);
+  DCHECK(success);
+
+  return next_payment_date.ToDoubleT();
+}
+
+uint64_t Balance::CalculateAdNotificationsReceivedThisMonth() const {
+  if (payments_.size() == 0) {
+    return 0;
+  }
+
+  PaymentInfo payment(payments_.front());
+  return payment.transaction_count;
+}
+
+bool Balance::HasPendingBalanceForLastMonth() const {
+  if (payments_.size() < 2) {
+    return false;
+  }
+
+  PaymentInfo payment(payments_.at(1));
+  if (payment.balance == 0.0) {
+    return false;
+  }
+
+  return true;
+}
+
+bool Balance::HasPendingBalanceForThisMonth() const {
+  if (payments_.size() < 1) {
+    return false;
+  }
+
+  PaymentInfo payment(payments_.front());
+  if (payment.balance == 0.0) {
```

Comment:
nit: could do `return payment.balance != 0.0` and we could reduce code a little bit

Reply:
We could but I find optimisations as such make the code less clear